FPRClient: fix paging API responses

mistydemeo opened the pull request:
Rather than try to fetch all items at once (which doesn't work for sets of results larger than 1000), rework the existing code into a simple fetch function and a generator that iterates over every item of the requested type. These are now fetched in batches of 100 and the "next" URL provided by the API response is used to page through all results.

Commit:
FPRClient: support paging API responses

Rather than try to fetch all items at once (which doesn't work for
sets of results larger than 1000), rework the existing code into a
simple fetch function and a generator that iterates over every item
of the requested type. These are now fetched in batches of 100 and
the "next" URL provided by the API response is used to page through
all results.

Fixes #6964.

Commit:
FPRClient: cleanups

* PEP8 cleanups
* dict comprehension instead of dict(list comprehension)

Hwesta commented on src/archivematicaCommon/lib/utilities/FPRClient/getFromRestAPI.py:
```diff
@@ -23,18 +23,37 @@
 
-def getFromRestAPI(url, resource, params, verbose=False, auth=None, verify=True):
+
+def getFromRestAPI(resource="", params=None, url="https://fpr.archivematica.org/fpr/api/v2/", limit=0, start_at=None, verbose=False, auth=None, verify=True):
```
Docstring for this? What the parameters are and how this will behave is non-obvious. (I'm thinking interactions between resource and url)

Hwesta commented on src/archivematicaCommon/lib/utilities/FPRClient/getFromRestAPI.py:
```diff
@@ -45,12 +64,31 @@ def getFromRestAPI(url, resource, params, verbose=False, auth=None, verify=True)
+            yield object
 
 if __name__ == '__main__':
     parser = OptionParser()
```
Update this section to use each_record instead of getFromRestAPI? Useful for testing.

mistydemeo replied:
Updated.

Hwesta commented:
Looks good!  Nice use of generator functions and urlparse (I hadn't heard of that library before).

Tests?  We don't have an easy way to specify vcrpy is a dependency, but it would work well here.

Out of scope for this, but do we want to move this out of utilities?  I'm not sure what benefit we get from keeping it there.

Commit:
getFromRestAPI: test each_record, not getFromRestAPI

refs #6964

Commit:
FPRClient: Mark getFromRestAPI function as private

Outside code should prefer using each_record() now.
Also rename to pothole_case.

refs #6964

mistydemeo commented:
Added verbose docstrings. Prefixed `getFromRestAPI` with an underscore to mark it as private now.

I'll look at adding some vcr.py tests.

mistydemeo commented:
> Out of scope for this, but do we want to move this out of utilities? I'm not sure what benefit we get from keeping it there.

I am not clear why it's there in the first place! Was it seen as a commandline utility at one point?

Commit:
FPRClient: update tests

* Add new test for each_record
* Adjust both each_record and getFromRestAPI tests to use VCR, and
  add recorded responses for both

refs #6964

Hwesta commented on src/archivematicaCommon/lib/utilities/FPRClient/getFromRestAPI.py:
```diff
@@ -23,18 +23,75 @@
 
-def getFromRestAPI(url, resource, params, verbose=False, auth=None, verify=True):
+
+def _getFromRestAPI(resource="", params=None, url="https://fpr.archivematica.org/fpr/api/v2/", limit=0, start_at=None, verbose=False, auth=None, verify=True):
```
Since you're renaming it anyway, you could make it potholes instead of camels. :)  Totally optional.

Hwesta commented:
I can only assume that it was intended for command line use, but I really don't know.  Historical reasons?

Commit:
FPRClient: Replace fprserver fixture with constant

Since this was unchanging, it can simply be a constant string without
breaking anything.

Since pytest fixtures don't always play nicely with VCR, it's easier
to sidestep this issue where possible.

refs #6964

mistydemeo commented:
Renamed `_get_from_rest_api` to pothole case. Adjusted tests: now uses VCR both for the old `_get_from_rest_api` test and for a new `each_record` test, and converted `fprserver` into a constant.

Hwesta commented:
:+1:

Commit:
Add testing requirements to requirements.txt

mistydemeo commented:
Added test requirements to requirements.txt in f1955fbbb53c86b9781e4f56cea248cb2769c07a.
